### ISSUE TITLE
Need benchmark for sentence selection task

### ISSUE THREAD

**Issue by pcgreat** [facebookresearch/ParlAI#213]
Just like DrQA becoming a high standard benchmark for span selection task for reading comprehension, we might also need a benchmark for sentence selection task (e.g. wikiqa). The ir_baseline is okay, just a little too simple.
I recently re-implemented the paper [A Compare-Aggregate Model for Matching Text Sequences](https://arxiv.org/abs/1611.01747) with PyTorch and ParlAI (author's original torch implementation is [here](https://github.com/shuohangwang/SeqMatchSeq)), which achieves 0.74 MAP for wikiqa. I can send a pull request if you would like to use it as an agent.

**Comment by jaseweston** [comment]
yes sounds fantastic !!!

**Comment by pcgreat** [comment]
Nice! I do have several questions when implementing it, and can I ask for suggestions here?
1. I would like to use SimpleDictionaryAgent which DrQA is using, so shall I copy it to my own agent folder, or put SimpleDictionaryAgent in the parlai/core/ folder
2. Similar, I need to use MAP and MRR as metrics to reproduce author's results, shall I create my own Metrics object, or just add them in parlai/core/metric.py?
3. I have a problem for building the dictionary while training the model. By default it only uses training data to build dictionary, and I wonder if there is a way to build dict with valid and test data. My current solution is to run a separate script to build dict with all train,valid, test data before the training, but I wonder if there is a better solution?

**Comment by jaseweston** [comment]
On Fri, Jul 14, 2017 at 1:24 PM, Chao Pan <notifications@github.com> wrote:

> Nice! I do have several questions when implementing it, and can I ask for
> suggestions here?
>
>    1. I would like to use SimpleDictionaryAgent which DrQA is using, so
>    shall I copy it to my own agent folder, or put SimpleDictionaryAgent in the
>    parlai/core/ folder
>
> can you not import it from the other agent? we could also put it as a
separate agent in parlai/agents ?


>
>    1. Similar, I need to use MAP and MRR as metrics to reproduce author's
>    results, shall I create my own Metrics object, or just add them in
>    parlai/core/metric.py?
>
>
I think add to the standard metrics class. We could maybe add a param to it
where you list which ones you want it to compute or something?

**Comment by Henry-E** [comment]
When I made a copy of the simple dictionary agent I followed the example of DrQA and put it into the same module as the [training agent class](https://github.com/Henry-E/ParlAI/blob/webnlg/parlai/agents/webnlg/webnlg.py). On a side note the simple dictionary agent option of only adding indices that appear in an embedding file is a nice option that might be worth adding to the default dictionary agent. 

Trying to build a dictionary using the validation and test sets was something I also ran into. It does look a little bit messy to do at the moment. The solution I went with works for my model because I only need to track out-of-vocabulary words for a copying mechanism, which is external to the network itself. So I [extend the dictionary](https://github.com/Henry-E/ParlAI/blob/webnlg/parlai/agents/webnlg/webnlg.py#L63)  with words as they appear inside the [build_example](https://github.com/Henry-E/ParlAI/blob/webnlg/parlai/agents/webnlg/webnlg.py#L181) function.

**Comment by pcgreat** [comment]
thanks @jaseweston nice suggestions! 

@Henry-E , thanks for your explanations and examples. If I understand correctly, you are trying to extend your dictionary when you see new data in validation/test (in act()), but the embeddings of these new words won't be updated, as embeddings are loaded during model initialization, right?

**Comment by Henry-E** [comment]
Yes the embeddings are initialised at the start and the new words remain UNKs inside the model. I guess I was mostly acknowledging that it was a problem I ran into as well. Though fortunately it was an easy enough fix for this particular model architecture.

**Comment by jaseweston** [comment]
Looks like this is dead? closing for now...
